lib/pmdtester/parsers/pmd_report_document.rb

```diff
@@ -0,0 +1,65 @@
+require 'nokogiri'
+require_relative '../pmd_violation'
+require_relative '../pmd_error'
+module PmdTester
+  # This class is used for registering types of events you are interested in handling.
+  # Also see: https://www.rubydoc.info/github/sparklemotion/nokogiri/Nokogiri/XML/SAX/Document
+  class PmdReportDocument < Nokogiri::XML::SAX::Document
+    def initialize(branch_name, violations, errors)
+      @violations = violations
+      @errors = errors
+      @current_violations = []
+      @current_violation = nil
+      @violations_size = 0
+      @current_error = nil
+      @errors_size = 0
+      @current_element = ''
+      @filename = ''
+      @branch_name = branch_name
+    end
+
+    def start_element(name, attrs = [])
+      @current_element = name
+
+      case name
+      when 'file'
+        @current_violations = []
+        @current_filename = attrs[0][1]
+      when 'violation'
+        @current_violation = PmdViolation.new(attrs, @branch_name)
+        @violations_size += 1
+      when 'error'
+        @current_filename = attrs[0][1]
+        @current_error = PmdError.new(attrs, @branch_name)
+        @errors_size += 1
+      end
+    end
+
+    def characters(string)
+      @current_violation.text = string unless @current_violation.nil?
+      @current_error.text = string unless @current_error.nil?
+    end
+
+    def end_element(name)
+      case name
+      when 'file'
+        @violations.violations.store(@current_filename, @current_violations)
```

Comment:
From object-oriented design, this violates the principle of encapsulation. E.g. PmdReportDocument now needs to know the inner structure of PmdViolations (that is stores the violations in a hash filename -> array of violations). You can create a method in PmdViolations, such that this line reads:

```
@violations.add_violations_by_filename(@current_filename, @current_violations)
```